### PR TITLE
chore(tracer): improve span sampling rules typing

### PR DESCRIPTION
**Pull request by ZStriker19**
## Description
Improve typing in SpanSampling and allow the setup of the rules to pass in the correct defaults as shown here: https://github.com/DataDog/dd-trace-py/pull/3977

## Checklist
- [ ] Title must conform to [conventional commit](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional).
- [ ] Add additional sections for `feat` and `fix` pull requests.
- [ ] Ensure tests are passing for affected code.
- [ ] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Relevant issue(s)
The typing allowing for the optional defaults was not as strict and helpful for understanding the class.

## Testing strategy
Tests were updated to reflect changes.

## Reviewer Checklist
- [x] Title is accurate.
- [x] Description motivates each change.
- [x] No unnecessary changes were introduced in this PR.
- [x] PR cannot be broken up into smaller PRs.
- [x] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [ ] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.
- [ ] Add to milestone.

### CODE REVIEW

File: ddtrace/internal/sampling.py [hunk 1]
```diff
@@ -112,18 +112,15 @@ class SpanSamplingRule:
...
+        sample_rate,  # type: float
+        max_per_second,  # type: int
         service=None,  # type: Optional[str]
         name=None,  # type: Optional[str]
-        sample_rate=1.0,  # type: Optional[float]
-        max_per_second=None,  # type: Optional[int]
     ):
         self.set_sample_rate(sample_rate)
         self._max_per_second = max_per_second
         # If no max_per_second specified then there is no limit
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/sampling.py:122]:
```suggestion
```

File: ddtrace/internal/sampling.py [hunk 2]
```diff
@@ -171,12 +169,15 @@ def match(self, span):
                 name_match = self._name_matcher.match(name)
         return service_match and name_match
 
-    def set_sample_rate(self, sample_rate=1.0):
-        self._sample_rate = float(sample_rate)
+    def set_sample_rate(self, sample_rate):
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/sampling.py:172]: seems like we don't really need this helper, it's only used once

File: ddtrace/internal/sampling.py [hunk 3]
```diff
@@ -112,18 +112,16 @@ class SpanSamplingRule:
 
     def __init__(
         self,
+        sample_rate,  # type: float
+        max_per_second,  # type: int
         service=None,  # type: Optional[str]
         name=None,  # type: Optional[str]
```
Comment (mabdinur) [inline ddtrace/internal/sampling.py:118]: Is there a case where `SpanSamplingRule` is initialized without the `service` and `name` arguments? Should we remove the default values?

**Review by mabdinur — APPROVED** [review]
Left a question that's unrelated to the typing change. LGTM

File: ddtrace/internal/sampling.py (hunk 3 above)
Comment (ZStriker19) [reply ddtrace/internal/sampling.py:118]: Yeah neither of service nor name are required, although there always has to be at least one.

**Review by brettlangdon — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

**Pull request merged**